### PR TITLE
Added JavaScript translations for List methods

### PR DESCRIPTION
**Pull request by acizmarik**
As suggested by #831, this PR adds various translations for methods on `List<T>`

### Regular `List<T>` methods

* `Add(element)`
* `AddRange(elements)`
* `Clear()`
* `Insert(index, element)`
* `InsertRange(index, elements)`
* `RemoveAt(index)`
* `RemoveAll(predicate)`
* `RemoveRange(index, count)`
* `Reverse()`

### Custom extensions
* `AddOrUpdate(element, matcher, updater)`
* `RemoveFirst(predicate)`
* `RemoveLast(predicate)`

### CODE REVIEW

File: src/DotVVM.Framework.Tests.Common/Binding/JavascriptCompilationTests.cs [hunk 1]
```diff
@@ -514,7 +557,7 @@ public void JsTranslator_EnumerableDistinct_NonPrimitiveTypesThrows(string bindi
         public void JsTranslator_EnumerableMax(string binding, string property)
         {
             var result = CompileBinding(binding, new[] { new NamespaceImport("System.Linq") }, new[] { typeof(TestArraysViewModel) });
-            Assert.AreEqual($"dotvvm.arrayHelper.max({property}(),function(arg){{return ko.unwrap(arg);}})", result);
+            Assert.AreEqual($"dotvvm.arrayHelper.max({property},function(arg){{return ko.unwrap(arg);}})", result);
```
Comment (exyi) [inline src/DotVVM.Framework.Tests.Common/Binding/JavascriptCompilationTests.cs:560]:
```suggestion
            Assert.AreEqual($"Math.max.apply(null, ({property}.map(function(arg){{return ko.unwrap(arg);}})", result);
```

Comment (acizmarik) [reply src/DotVVM.Framework.Tests.Common/Binding/JavascriptCompilationTests.cs]: `Math.max` behaves differently on empty collection

**Comment by acizmarik** [comment]
Fixed issues with Enumerable.Min, Enumerable.Max to behave correctly on nullable types. Added tests for `arrayHelper` JS functions.

File: src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts [hunk 2]
```diff
@@ -1,121 +1,215 @@
...
+function min<T>(array: T[], selector: (item: T) => number, throwIfEmpty: boolean): number | null {
+    if (array.length === 0) {
+        if (throwIfEmpty) {
+            throw new Error("Source is empty! Min operation cannot be performed.");
+        }
+        return null;
+    }
+
+    if (array.length == 1)
+        return selector(array[0]);
```
Comment (exyi) [inline src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts:153]:
```suggestion
```

File: src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts [hunk 3]
```diff
@@ -1,121 +1,215 @@
...
+function max<T>(array: T[], selector: (item: T) => number, throwIfEmpty: boolean): number | null {
+    if (array.length === 0) {
+        if (throwIfEmpty) {
+            throw new Error("Source is empty! Max operation cannot be performed.");
+        }
+        return null;
+    }
+
+    if (array.length == 1)
+        return selector(array[0]);
```
Comment (exyi) [inline src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts:134]:
```suggestion
```

File: src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts [hunk 4]
```diff
@@ -1,121 +1,215 @@
...
-            if (r[j] == source[i]) {
+            if (r[j] == array[i]) {
                 found = true;
                 break;
             }
         }
         if (found)
             continue;
-        r.push(source[i]);
+        r.push(array[i]);
```
Comment (exyi) [inline src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts:90]: you can do just `return Array.from(new Set(array))`, maybe `return Array.from(new Set(array.map(ko.unwrap)))`, not sure.

File: src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts [hunk 5]
```diff
@@ -1,121 +1,215 @@
...
+    }
+
+    observable.setState(array);
+}
+
+function any<T>(array: T[], predicate: (s: T) => boolean): boolean {
+    return firstOrDefault(array, predicate) != null;
+}
+
+function all<T>(array: T[], predicate: (s: T) => boolean): boolean {
```
Comment (exyi) [inline src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts:65]: Is there a difference to `array.every(predicate)`?

File: src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts [hunk 6]
```diff
@@ -1,121 +1,215 @@
...
+function addRange<T>(observable: any, elements: T[]): void {
+    let array = Array.from<T>(observable.state);
+    for (let i = 0; i < elements.length; i++) {
+        array.push(ko.unwrap(elements[i]));
+    }
+
+    observable.setState(array);
+}
+
+function any<T>(array: T[], predicate: (s: T) => boolean): boolean {
```
Comment (exyi) [inline src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts:61]: Is there a difference to `array.some(predicate)`?

**Review by exyi — APPROVED** [review]
Just a few last suggestions

File: src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts (hunk 4 above)
Comment (acizmarik) [reply src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts]: The second option works

File: src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts (hunk 5 above)
Comment (acizmarik) [reply src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts]: Probably not. I was just not aware that this function already exists

**Pull request merged**